src/wallet/wallet.cpp

```diff
@@ -1935,29 +1935,58 @@ OutputType CWallet::TransactionChangeType(const std::optional<OutputType>& chang
         return *change_type;
     }
 
-    // if m_default_address_type is legacy, use legacy address as change (even
-    // if some of the outputs are P2WPKH or P2WSH).
+    // if m_default_address_type is legacy, use legacy address as change.
     if (m_default_address_type == OutputType::LEGACY) {
         return OutputType::LEGACY;
     }
 
-    // if any destination is P2WPKH or P2WSH, use P2WPKH for the change
-    // output.
+    bool any_tr{false};
+    bool any_wpkh{false};
+    bool any_sh{false};
+    bool any_pkh{false};
+
     for (const auto& recipient : vecSend) {
-        // Check if any destination contains a witness program:
-        int witnessversion = 0;
-        std::vector<unsigned char> witnessprogram;
-        if (recipient.scriptPubKey.IsWitnessProgram(witnessversion, witnessprogram)) {
-            if (GetScriptPubKeyMan(OutputType::BECH32M, true)) {
-                return OutputType::BECH32M;
-            } else if (GetScriptPubKeyMan(OutputType::BECH32, true)) {
-                return OutputType::BECH32;
-            } else {
-                return m_default_address_type;
-            }
-        }
+        std::vector<std::vector<uint8_t>> dummy;
+        const TxoutType type{Solver(recipient.scriptPubKey, dummy)};
+        if (type == TxoutType::WITNESS_V1_TAPROOT) {
```

Comment:
Unknown witness versions should be included here too.

Reply:
I'd say no. Otherwise, it will be trivial for chain-analysis to determine the change type if the outputs are {v2, v1, v0} and the input is {v1}

Reply:
Yes, but the intention is to use `OutputType::BECH32M` for future witness versions too. v2 outputs will have bech32m addresses. If you had a v2 descriptor as active, it would take the slot that tr descriptors would currently occupy so a tx with a taproot output would have a v2 as change.

Reply:
This is impossible to do unless the specification of v2 is known.

If I use UNKNOWN here and V2 is specified, then UNKNOWN will *not* match V2. Before V2 is defined, it is not possible to match it. As soon as V2 is defined, it can be matched. So this is something for a follow-up when the time is right.

Reply:
Sure, but I think we should handle UNKNOWN here too as one of the explicitly stated goals of bech32(m) is to allow sending to future segwit versions without needing to know about them. When a new version's TxoutType is added, it needs to be added here as well, but we should always handle UNKNOWN regardless.

Reply:
UNKNOWN is handled. Obviously it is impossible to match, but it is treated like any other type that can't be matched (wsh, pk, raw multisig, raw data, ...).

Happy to add a comment, but the algorithm is:

* find any types that can theoretically be matched
* try to match them if there is a spkman
* fallback to bech32(m) (Note: This will happen if *all* outputs are unmatchable. For example all are v2)

Reply:
Not sure what to do here, but I think you'll have to send me a sketch of a diff to clarify what you mean.

Reply:
I was thinking something like

```suggestion
        if (type == TxoutType::WITNESS_V1_TAPROOT || type == TxoutType::WITNESS_UNKNOWN) {
```

But looking at this further, I suppose the current implementation is fine. If the transaction contains any known address types, then we will match to those and privacy is still preserved. If they are all unknown, then we will try to use bech32(m) anyways, unless unavailable, in which case this wouldn't matter.

Reply:
Right, on mismatch it'd try bech32(m)